PTC-59 How many hours did I work in total until today

dsp opened the pull request.

dsp committed:
Retreive booked minutes by start and end date

We generalize the getBookedMinutes method and use a generic start and
end date. We use this method in minutesBookedAction.

dsp committed:
Make the end date in the model api exclusive

We choosed to use exclusive end dates therefore I changed the model API
to use exclusive end dates.

dsp committed:
Deprecate legacy jsonMonthListAction() in favor of monthListAction()

dsp committed:
Use start and end dates in monthJsonList

dsp committed:
Remove unused code and refactor gteRecords

We don't need openPerid & co in the frontend therefore we don't
need the code to determine if a period is open or not. We can
reduce getRecords to dates and a sum that can be easily expressed
in SQL.

dsp committed:
Use timecard model in SummaryBlock

dsp committed:
Use start and end dates for timeard calls in frontend

dsp committed:
If no start date is given use last contracts start date

dsp committed:
Reuse fetchByUserAndPeriod

dsp committed:
Javascript

dsp committed:
Use a table to display summaries

dsp committed:
Remove unused function

dsp committed:
End date is always today

dsp committed:
Merge branch 'next' into PTC-59

Conflicts:
	phprojekt/application/Timecard/Controllers/IndexController.php
	phprojekt/application/Timecard/Models/Timecard.php
	phprojekt/htdocs/phpr/BookingList/SummaryBlock.js
	phprojekt/htdocs/phpr/models/Timecard.js

dsp committed:
Use today as default endday

The current calculations for the current month use the whole month it's
calculation period. This results in a "hours to work" to be calculated
for the full month. In fact the user is interested how much he has to
work till today. So we change the default end date to today.

exi commented on phprojekt/application/Timecard/Controllers/IndexController.php:
```diff
@@ -47,13 +47,13 @@ public function setCurrentProjectId()
-        $month   = (int) $this->getRequest()->getParam('month', date("m"));
-        $records = Timecard_Models_Timecard::getMonthRecords($year, $month);
+        list($start, $end) = $this->_paramToStartEndDT();
+        $records = $this->getModelObject()->getRecords($start, $end);
```
dont' user getModelObject anymore, look at line 67

exi replied:
indentation

dsp replied:
done

exi commented on phprojekt/application/Timecard/Controllers/IndexController.php:
```diff
@@ -63,10 +63,8 @@ public function jsonMonthListAction()
-        $projects = $this->_projectsParamToArray();
-        $records = Timecard_Models_Timecard::getMonthRecords($year, $month, $projects);
+        list($start, $end) = $this->_paramToStartEndDT();
+        $records = Timecard_Models_Timecard::getRecords($start, $end);
```
indentation

dsp replied:
done

exi commented on phprojekt/application/Timecard/Controllers/IndexController.php:
```diff
@@ -328,11 +326,8 @@ public function csvListAction()
-
-        $minutes = Timecard_Models_Timecard::getBookedMinutesInMonth($year, $month, $projects);
+        list($start, $end) = $this->_paramToStartEndDT();
+        $minutes = Timecard_Models_Timecard::getBookedMinutes($start, $end);
```
indentation

dsp replied:
done

exi commented on phprojekt/application/Timecard/Controllers/IndexController.php:
```diff
@@ -388,9 +383,9 @@ public function setParams()
+        list($start, $end) = $this->_paramToStartEndDT();
 
-        $contracts = Timecard_Models_Contract::fetchByUserAndPeriod(Phprojekt_Auth::getRealUser(), $start, $end);
+        $contracts     = Timecard_Models_Contract::fetchByUserAndPeriod(Phprojekt_Auth::getRealUser(), $start, $end);
```
don't use getRealUser but Phprojekt_Auth_Proxy::getEffectiveUser

dsp replied:
done

exi commented on phprojekt/application/Timecard/Controllers/IndexController.php:
```diff
@@ -458,18 +453,17 @@ public function projectUserMinutesAction()
+        $start = $this->getRequest()->getParam('start');
+
+        /* either given start day or the start day of the contract. */
+        $first = array_pop(Timecard_Models_Contract::fetchByUser(Phprojekt_Auth::getRealUser()));
```
don't use getRealUser but Phprojekt_Auth_Proxy::getEffectiveUser

dsp replied:
done

dsp committed:
Use static method call

dsp committed:
Fix indention

dsp committed:
Use getEffectiveUser() instead of getRealUser()

We should always user getEffectiveUser() instead of getRealUser() to make
sure our backend is proxy mode compatible.

exi commented on phprojekt/application/Timecard/Models/Contract.php:
```diff
@@ -19,19 +19,32 @@
 
+    public static function fetchByUser(Phprojekt_User_User $user, $limit = null)
+    {
+        return self::fetchByUserAndPeriod($user, null, null);
```
why not set proper default values for the last 2 parameters instead of passing magic 'null's?

dsp replied:
Because I don't want any where limitations in the query.

dsp committed:
$start and $end are optional parameters in fetchByUserAndPeriod

dsp committed:
Readd project filter but make it optional

exi commented on phprojekt/htdocs/phpr/BookingList/SummaryBlock.js:
```diff
@@ -9,8 +9,9 @@ define([
+    'phpr/models/Timecard',
     'dojo/text!phpr/template/bookingList/summaryBlock.html'
-], function(declare, lang, topic, domStyle, all, DeferredList, _WidgetBase, _TemplatedMixin, api, timehelper, templateString) {
+], function(declare, lang, topic, domStyle, all, DeferredList, _WidgetBase, _TemplatedMixin, api, timehelper, timecardModel, templateString) {
```
please make the list 'flat' if you touch it, we'll gradually convert all of these to a one-per-line style to ease diffing

exi commented on phprojekt/htdocs/phpr/BookingList/SummaryBlock.js:
```diff
@@ -48,33 +49,24 @@ define([
         _update: function() {
+            timecardModel.getMinutesBookedTotal()
+            .then(lang.hitch(this, function(data) {
+                this.bookedTotal.innerHTML = timehelper.minutesToHMString(data.minutesBooked);
```
check for this._destroyed after async calls, the dom node might be gone by now

exi commented on phprojekt/htdocs/phpr/BookingList/SummaryBlock.js:
```diff
@@ -48,33 +49,24 @@ define([
+
+            timecardModel.getMinutesToWorkTotal()
+            .then(lang.hitch(this, function(data) {
+                this.toWorkTotal.innerHTML = timehelper.minutesToHMString(data.minutesToWork);
```
check for this._destroyed after async calls, the dom node might be gone by now

exi commented on phprojekt/htdocs/phpr/BookingList/SummaryBlock.js:
```diff
@@ -48,33 +49,24 @@ define([
-            ).then(lang.hitch(this, function(data) {
+            var bookedPromise = timecardModel.getMinutesBooked()
+            .then(lang.hitch(this, function(data) {
                 this.booked.innerHTML = timehelper.minutesToHMString(data.minutesBooked);
```
check for this._destroyed after async calls, the dom node might be gone by now

exi commented on phprojekt/htdocs/phpr/BookingList/SummaryBlock.js:
```diff
@@ -48,33 +49,24 @@ define([
-                domStyle.set(this.toWork, 'display', 'none');
+            var toWorkPromise = timecardModel.getMinutesToWork()
+            .then(lang.hitch(this, function(data) {
+                this.toWork.innerHTML = timehelper.minutesToHMString(data.minutesToWork);
```
check for this._destroyed after async calls, the dom node might be gone by now

dsp committed:
Include days without bookings in resultsset

dsp committed:
Fix wrong method name

dsp committed:
Check if node is destroyed in async calls

dsp committed:
One-line imports

exi commented on phprojekt/application/Timecard/Models/Contract.php:
```diff
@@ -19,19 +19,32 @@
 class Timecard_Models_Contract extends Phprojekt_ActiveRecord_Abstract
 {
 
+    public static function fetchByUser(Phprojekt_User_User $user, $limit = null)
```
is the limit used anywhere?

exi commented on phprojekt/application/Timecard/Models/Timecard.php:
```diff
@@ -557,13 +539,10 @@ public function fetchAll($where = null, $order = null, $count = null, $offset =
+     * @param DateTime $end End date (excluded)
      **/
-    public static function getBookedMinutesInMonth($year, $month, Array $projects = null)
+    public static function getBookedMinutes(DateTime $start, DateTime $endExcluded)
```
projects filter still missing

dsp committed:
Readd project filter

dsp committed:
Remove unused limit

dsp committed:
Add a contract for unit tests

dsp committed:
Make sure we use a temporary variabale to be passed by references

dsp committed:
Unit tests should reflect the new monthList API

exi commented:
This pull request breaks the statistics table!

dsp committed:
Return records as array

exi commented on phprojekt/application/Timecard/Controllers/IndexController.php:
```diff
@@ -47,13 +47,14 @@ public function setCurrentProjectId()
-        $records = Timecard_Models_Timecard::getMonthRecords($year, $month);
+        list($start, $end) = $this->_paramToStartEndDT();
+        $projects          = $this->_projectsParamToArray();
+        $records           = Timecard_Models_Timecard::getRecords($start, $end, $projects);
```
jsonMonthList is only used by the legacy timecard which does not work anymore anyway, could we not just remove this action?

dsp replied:
will be removed in an upcoming pr

exi commented on phprojekt/application/Timecard/Controllers/IndexController.php:
```diff
@@ -63,10 +64,9 @@ public function jsonMonthListAction()
      */
     public function monthListAction()
```
I would rename this action to dateRangeAction because 'monthList' does not apply anymore if we use date ranges as the parameters.
I would rename the 'getMonthList' function in the frontend timecard model to match the new name too

dsp replied:
done

exi commented on phprojekt/application/Timecard/Models/Timecard.php:
```diff
@@ -299,19 +299,27 @@ public function getError()
      * @return array
      */
-    static public function getMonthRecords($year, $month, Array $projects = null)
+    public static function getRecords(\DateTime $start, \DateTime $end, $projects = null)
```
Why was the Array type hint for $projects removed?

dsp replied:
readded

exi commented on phprojekt/application/Timecard/Models/Timecard.php:
```diff
@@ -319,46 +327,20 @@ static public function getMonthRecords($year, $month, Array $projects = null)
+        $records  = [];
+        foreach(new DatePeriod($start, $interval, $end) as $date) {
+            $formated           = $date->format('Y-m-d');
+            $records[$formated] = ['date' => $formated, 'sumInMinutes' => 0];
```
why use $records[$formated] here?
$records[] = would to the same thing and you could remove the array_values() call in line 343

dsp replied:
the $records[$formated] is there to overwrite the values afterwards with what we got from the DB.

exi replied:
Ah, I see, ok, makes sense.

exi commented on phprojekt/application/Timecard/Models/Timecard.php:
```diff
@@ -557,13 +539,10 @@ public function fetchAll($where = null, $order = null, $count = null, $offset =
+     * @param DateTime $end End date (excluded)
      **/
-    public static function getBookedMinutesInMonth($year, $month, Array $projects = null)
+    public static function getBookedMinutes(DateTime $start, DateTime $endExcluded, $projects = null)
```
You used \DateTime as a type hint in line 302, here 'DateTime' which one is correct?

dsp replied:
in our case, both as we dont use namespaces anyway.

exi replied:
Would be nice if we would be consistent then

exi commented on phprojekt/application/Timecard/Models/Timecard.php:
```diff
@@ -581,6 +560,7 @@ public static function getBookedMinutesInMonth($year, $month, Array $projects =
         }
 
         return (int) $select->query()->fetchColumn();
+
```
newline after return serves no purpose

dsp replied:
done

dsp committed:
Typehint project array

dsp committed:
Use correct namespace when accessing DateTime

dsp committed:
Remove unecessary newline

dsp committed:
Rename monthList to workedMinutesPerPday

dsp committed:
JS Checkstyle
